File: frontend/features/prompt-studio/templating.ts
/**
 * ${name} placeholder substitution - the TypeScript mirror of
 * backend/app/services/prompt_studio/templating.py. Rendering-determinism
 * constraint (01_SPEC.md §3.16, project-owner requirement): this module
 * performs substitution, escaping, and validation only - never a conditional,
 * loop, function call, filter, include, or import. Never replace this with a
 * templating library (Handlebars, Mustache, etc.); that would be a new
 * architectural decision, not an incremental change here.
 */

const PLACEHOLDER_PATTERN = /\$\$|\$([A-Za-z_][A-Za-z0-9_]*)|\$\{([A-Za-z_][A-Za-z0-9_]*)\}/g;

/** Every distinct ${name} referenced in body, using the same pattern substitute() below scans with. */
export function extractPlaceholders(body: string): Set<string> {
  const names = new Set<string>();
  for (const match of body.matchAll(PLACEHOLDER_PATTERN)) {
    const name = match[1] ?? match[2];
    if (name) names.add(name);
  }
  return names;
}

export interface SubstituteResult {
  rendered: string;
  missing: string[];
}

/**
 * Best-effort client-side preview substitution. Unlike the backend's strict
 * substitute() (which raises on a missing key), this always returns a
 * renderable string - a missing/blank required variable keeps its ${name}
 * token visible in the output and is also reported in `missing`, so the
 * caller can highlight it as unresolved rather than silently rendering an
 * empty string (01_SPEC.md §3.5).
 */
export function substitute(body: string, values: Record<string, string>): SubstituteResult {
  const missing = new Set<string>();
  const rendered = body.replace(PLACEHOLDER_PATTERN, (match, named?: string, braced?: string) => {
    if (match === "$$") return "$";
    const name = named ?? braced;
    if (!name) return match;
    const value = values[name];
    if (value === undefined || value === "") {
      missing.add(name);
      return `\${${name}}`;
    }
    return value;
  });
  return { rendered, missing: Array.from(missing) };
}

File: backend/app/services/prompt_studio/templating.py
"""Deterministic ${name} placeholder substitution for prompt bodies.

Rendering-determinism constraint (project-owner requirement, spec review
2026-07-22 - see forge-docs/implementation/Phase-03-Prompt-Studio/01_SPEC.md
§3.16): this module performs substitution, escaping, and validation only.
It must never grow a conditional, loop, function call, filter, include, or
import - string.Template is used specifically because it has no expression
language at all, so that guarantee is a property of the tool, not just a
policy on top of it. Do not replace this with Jinja2 or any other templating
engine; that would be a new architectural decision, not an incremental change
here.
"""

from __future__ import annotations

from string import Template


def extract_placeholders(body: str) -> set[str]:
    """Every distinct ${name} referenced in body, using string.Template's own
    pattern so extraction can never disagree with substitution."""
    names: set[str] = set()
    for match in Template.pattern.finditer(body):
        name = match.group("named") or match.group("braced")
        if name:
            names.add(name)
    return names


def substitute(body: str, values: dict[str, object]) -> str:
    """Strict substitution - raises KeyError via string.Template if a
    referenced name is missing from values. Callers that need a
    best-effort/partial render should catch that and surface the missing
    names, not silently render empty strings."""
    return Template(body).substitute(values)
